src/layer/lstm.h

```diff
@@ -38,6 +39,11 @@ class LSTM : public Layer
     Mat weight_hc_data;
     Mat weight_xc_data;
     Mat bias_c_data;
+
+    mutable Mat hidden;
+    mutable Mat cell;
```

Comment:
Declaring mutable members in layer and trying to modify it in forward() function breaks thread safety.
To keep the state, hidden and cell shall be retrieved from bottom_blobs and returned in top_blobs.

Reply:
Hmm I see, passing and retrieving hidden states via bottom and top blobs will be quite a hassle once there are multiple layers involved. Do you think there would be some other options where the state gets stored in the Net instance instead of expecting the user to pass and retrieve the hidden states?

Reply:
I suggest adding the hidden / cell blobs like how onnx does
https://github.com/onnx/onnx/blob/master/docs/Operators.md#LSTM
initial_h initial_c
Y_h Y_c

The data race happens when two Extractor instances of the same LSTM model are running concurrently. Weight data can be shared as they are always readonly.

Reply:
I see i had not considered that use case, in that case i will have to propagate the state blobs myself. I will rework the pr to include a state input and output.